cpp/include/raft/spectral/spectral_util.hpp

```diff
@@ -0,0 +1,235 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/cudart_utils.h>
+#include <raft/handle.hpp>
+#include <raft/spectral/sm_utils.hpp>
+
+#include <thrust/device_vector.h>
+#include <thrust/fill.h>
+#include <thrust/reduce.h>
+#include <thrust/transform.h>
+
+namespace raft {
+namespace spectral {
+
+template <typename IndexType_, typename ValueType_>
+static __global__ void scale_obs_kernel(IndexType_ m, IndexType_ n,
+                                        ValueType_* obs) {
+  IndexType_ i, j, k, index, mm;
+  ValueType_ alpha, v, last;
+  bool valid;
+  // ASSUMPTION: kernel is launched with either 2, 4, 8, 16 or 32 threads in x-dimension
+
+  // compute alpha
+  mm = (((m + blockDim.x - 1) / blockDim.x) *
+        blockDim.x);  // m in multiple of blockDim.x
+  alpha = 0.0;
+  // printf("[%d,%d,%d,%d] n=%d, li=%d, mn=%d \n",threadIdx.x,threadIdx.y,blockIdx.x,blockIdx.y, n,
+  // li, mn);
+  for (j = threadIdx.y + blockIdx.y * blockDim.y; j < n;
+       j += blockDim.y * gridDim.y) {
+    for (i = threadIdx.x; i < mm; i += blockDim.x) {
+      // check if the thread is valid
+      valid = i < m;
+
+      // get the value of the last thread
+      last = utils::shfl(alpha, blockDim.x - 1, blockDim.x);
+
+      // if you are valid read the value from memory, otherwise set your value to 0
+      alpha = (valid) ? obs[i + j * m] : 0.0;
+      alpha = alpha * alpha;
+
+      // do prefix sum (of size warpSize=blockDim.x =< 32)
+      for (k = 1; k < blockDim.x; k *= 2) {
+        v = utils::shfl_up(alpha, k, blockDim.x);
+        if (threadIdx.x >= k) alpha += v;
+      }
+      // shift by last
+      alpha += last;
+    }
+  }
+
+  // scale by alpha
+  alpha = utils::shfl(alpha, blockDim.x - 1, blockDim.x);
+  alpha = std::sqrt(alpha);
+  for (j = threadIdx.y + blockIdx.y * blockDim.y; j < n;
+       j += blockDim.y * gridDim.y) {
+    for (i = threadIdx.x; i < m; i += blockDim.x) {  // blockDim.x=32
+      index = i + j * m;
+      obs[index] = obs[index] / alpha;
+    }
+  }
+}
+
+template <typename IndexType_>
+IndexType_ next_pow2(IndexType_ n) {
+  IndexType_ v;
+  // Reference:
+  // http://graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2Float
+  v = n - 1;
+  v |= v >> 1;
+  v |= v >> 2;
+  v |= v >> 4;
+  v |= v >> 8;
+  v |= v >> 16;
+  return v + 1;
+}
+
+template <typename IndexType_, typename ValueType_>
+cudaError_t scale_obs(IndexType_ m, IndexType_ n, ValueType_* obs) {
+  IndexType_ p2m;
+  dim3 nthreads, nblocks;
+
+  // find next power of 2
+  p2m = next_pow2<IndexType_>(m);
+  // setup launch configuration
+  nthreads.x = max(2, min(p2m, 32));
```

Comment:
Why not use the `dim3` constructor on one-line?

Reply:
Because:

1. there are `narrowing conversion` warnings due to `dim3{}` constructor expecting the unsigned version of whatever index_type is; and doing an explicit cast to `unsigned index_type` would fail when / if `index_type` _is_ unisgned already;
2. `nthreads.y` is calculated in terms of `nthreads.x`, which would not work if these become constructor arguments;

Reply:
OK, fixed.